runelite-client/src/main/java/net/runelite/client/graphics/ModelOutlineRenderer.java

```diff
@@ -0,0 +1,1073 @@
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package net.runelite.client.graphics;
+
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import java.awt.Color;
+import java.awt.image.BufferedImage;
+import java.awt.image.DataBufferInt;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import net.runelite.api.Client;
+import net.runelite.api.DecorativeObject;
+import net.runelite.api.GameObject;
+import net.runelite.api.GroundObject;
+import net.runelite.api.MainBufferProvider;
+import net.runelite.api.Model;
+import net.runelite.api.NPC;
+import net.runelite.api.NPCComposition;
+import net.runelite.api.Perspective;
+import net.runelite.api.Player;
+import net.runelite.api.Projectile;
+import net.runelite.api.coords.LocalPoint;
+import net.runelite.client.task.Schedule;
+import net.runelite.client.task.Scheduler;
+
+@Singleton
+public class ModelOutlineRenderer
+{
+	/*
+	 * This class doesn't really "need" static variables, but they are
+	 * static for performance reasons. Arrays are kept outside methods
+	 * to avoid frequent big allocations. Arrays should mostly be seen
+	 * as ArrayLists. The size of them is increased whenever they need
+	 * to become bigger.
+	 */
+
+	@Inject
+	private Client client;
+
+	@Inject
+	private Scheduler scheduler;
+
+	private boolean isReset;
+	private boolean usedSinceLastCheck;
+
+	// Dimensions of the underlying image
+	private int imageWidth;
+	private int imageHeight;
+
+	// Boundaries for the current rasterization
+	private int clipX1;
+	private int clipY1;
+	private int clipX2;
+	private int clipY2;
+
+	// Pixel points that would be rendered to
+	private int[] visited;
+	private int currentVisitedNumber = 0;
+
+	// Transformed vertex positions
+	private int[] projectedVerticesX;
+	private int[] projectedVerticesY;
+	private boolean[] projectedVerticesRenderable;
+
+	// An array of pixel points to raster onto the image. These are checked against
+	// clip boundaries and the visited array to prevent drawing on top of the model
+	// and outside the scene area. They are grouped per distance to the closest pixel
+	// drawn on the model.
+	private int[][] outlinePixels;
+	private int[] outlinePixelsLengths; // outlinePixelsLength[i] is the used length of outlinePixels[i]
+	private int outlineArrayWidth;
+
+	// A list of pixel distances ordered from shortest to longest distance for
+	// each outline width. These are calculated once upon first usage and then
+	// stored here to prevent reevaluation.
+	private List<List<PixelDistanceAlpha>> precomputedDistancePriorities;
+
+	public ModelOutlineRenderer()
```

Comment:
Make this constructor private and mark it with `@Inject` (also you should probably inject both client and scheduler here in constructor).